### ISSUE TITLE
vmss: Adding an IpConfiguration using a jsonString returns a msrest serialization error

### ISSUE THREAD

**Issue by denisbollinne** [Azure/azure-cli#3321]
### Description
Hello,

I'm trying to add an IpConfiguration to a VMSS using the cli (using the docker image).
using the following command:
`az vmss update --name 'vmssName' --resource-group 'resourceGroupName' --add virtualMachineProfile.networkProfile.networkInterfaceConfigurations[0].ipConfigurations "$jsonData"`

`$jsonData` is set with the following json:
(i also tried to use double quotes instead of single quotes, but same result)
```
{
    'name':  'nic-internal',
    'subnet':  {
                   'id':  '/subscriptions/subscriptionId/resourceGroups/opisense-backend/providers/Microsoft.Network/virtualNetworks/backend-vnet/subnets/loadbalancer-subnet',
				   'resourceGroup': 'resourceGroupName'
               },
    'applicationGatewayBackendAddressPools':  null,
    'loadBalancerBackendAddressPools':  [
                                            {
                                                'id':  '/subscriptions/subscriptionId/resourceGroups/opisense-backend/providers/Microsoft.Network/loadBalancers/LB-cluster-internal/backendAddressPools/LB-backend',
												'resourceGroup': 'resourceGroupName'
                                            }
                                        ],
    'loadBalancerInboundNatPools':  [
                                        {
                                            'id':  '/subscriptions/subscriptionId/resourceGroups/opisense-backend/providers/Microsoft.Network/loadBalancers/LB-cluster-internal/inboundNatPools/LoadBalancerBEAddressNatPool',
											'resourceGroup': 'resourceGroupName'
                                        }
                                    ],
    'id':  null
}
```
Then i'm getting a json (de)serialization error
```

'loadBalancerInboundNatPools'
Traceback (most recent call last):
  File "/usr/local/lib/python3.5/site-packages/azure/cli/main.py", line 36, in main
    cmd_result = APPLICATION.execute(args)
  File "/usr/local/lib/python3.5/site-packages/azure/cli/core/application.py", line 203, in execute
    result = expanded_arg.func(params)
  File "/usr/local/lib/python3.5/site-packages/azure/cli/core/commands/__init__.py", line 278, in __call__
    return self.handler(*args, **kwargs)
  File "/usr/local/lib/python3.5/site-packages/azure/cli/core/commands/arm.py", line 349, in handler
    raise ex
  File "/usr/local/lib/python3.5/site-packages/azure/cli/core/commands/arm.py", line 332, in handler
    opres = setter(client, **setterargs) if client else setter(**setterargs)
  File "/usr/local/lib/python3.5/site-packages/azure/cli/command_modules/vm/custom.py", line 1496, in vmss_set
    resource_group_name, name, raw=no_wait, **kwargs)
  File "/usr/local/lib/python3.5/site-packages/azure/mgmt/compute/compute/v2016_04_30_preview/operations/virtual_machine_scale_sets_operations.py", line 85, in create_or_update
    body_content = self._serialize.body(parameters, 'VirtualMachineScaleSet')
  File "/usr/local/lib/python3.5/site-packages/msrest/serialization.py", line 301, in body
    data = _convert_to_datatype(data, data_type, self.dependencies)
  File "/usr/local/lib/python3.5/site-packages/msrest/serialization.py", line 182, in _convert_to_datatype
    getattr(data, attr), mapconfig['type'], localtypes))
  File "/usr/local/lib/python3.5/site-packages/msrest/serialization.py", line 182, in _convert_to_datatype
    getattr(data, attr), mapconfig['type'], localtypes))
  File "/usr/local/lib/python3.5/site-packages/msrest/serialization.py", line 182, in _convert_to_datatype
    getattr(data, attr), mapconfig['type'], localtypes))
  File "/usr/local/lib/python3.5/site-packages/msrest/serialization.py", line 156, in _convert_to_datatype
    param, data_type[1:-1], localtypes) for param in data
  File "/usr/local/lib/python3.5/site-packages/msrest/serialization.py", line 156, in <listcomp>
    param, data_type[1:-1], localtypes) for param in data
  File "/usr/local/lib/python3.5/site-packages/msrest/serialization.py", line 182, in _convert_to_datatype
    getattr(data, attr), mapconfig['type'], localtypes))
  File "/usr/local/lib/python3.5/site-packages/msrest/serialization.py", line 156, in _convert_to_datatype
    param, data_type[1:-1], localtypes) for param in data
  File "/usr/local/lib/python3.5/site-packages/msrest/serialization.py", line 156, in <listcomp>
    param, data_type[1:-1], localtypes) for param in data
  File "/usr/local/lib/python3.5/site-packages/msrest/serialization.py", line 171, in _convert_to_datatype
    localtypes) for key in data
  File "/usr/local/lib/python3.5/site-packages/msrest/serialization.py", line 171, in <dictcomp>
    localtypes) for key in data
KeyError: 'loadBalancerInboundNatPools'
```

Would you have any idea what's wrong? is the json format invalid or is it simply not supported by the cli?

Thank you,

---

### Environment summary

**Install Method:** How did you install the CLI? (e.g. pip, interactive script, apt-get, Docker, MSI, nightly)  
Answer here: Docker (latest, image id 9521494feb6e)

**CLI Version:** What version of the CLI and modules are installed? (Use `az --version`)  
Answer here: 2.0.6

**OS Version:** What OS and version are you using?  
Answer here: Windows 10, but using Azure CLI inside docker 

**Shell Type:** What shell are you using? (e.g. bash, cmd.exe, Bash on Windows)  
Answer here: bash (from docker container)

**Comment by tjprescott** [comment]
Hi @denisbollinne, I would recommend you try a couple things:
1. Remove any properties that are null. Omission itself should accomplish that.
2. Change your keys from "camelCasing" to "snake_casing".  All properties in the SDK use snake casing, which may be why you are getting the KeyError.

**Comment by denisbollinne** [comment]
Thanks @tjprescott,

I applied your recommendations, and in addition, I had to remove the `resource_group` elements.
With all these changes, the json was parsed successfully.

But then my journey was not over, so for sake of completeness, here are the different steps i had to do in order to "Add a second IpConfiguration to an existing virtual scale set"

Closing the ticket.
---

1.  I had to use the command `az resource update` instead of `az vmss update` in order to use the latest API version (required to be able to add a second IpConfiguration).
1.1. using `az resource update` required the json to be changed, because the resources are nested in 
`properties` elements
2. I had to flag the existing IpConfiguration as 'Primary', this could be achieved with the following command `az resource update --id '/subscriptions/subscriptionId/resourceGroups/resourceGroupName/providers/Microsoft.Compute/virtualMachineScaleSets/vmssName'  --resource-group 'resourceGroupName' --set properties.virtualMachineProfile.networkProfile.networkInterfaceConfigurations[0].properties.ipConfigurations[0].properties primary=true`
3. Then i tried to use `az resource update`, but this time it complained with `subscription is not registered for feature Microsoft.Network/AllowPublicIpsAndLoadBalancingOnSecondaryIpConfigsInVMScaleSet required to carry out the requested operation.`
4. I could add the feature with this command  `az feature register --name 'AllowPublicIpsAndLoadBalancingOnSecondaryIpConfigsInVMScaleSet' --namespace 'Microsoft.Network'`

Thanks,
